pkg/util/validation/limits.go

```diff
@@ -0,0 +1,226 @@
+package validation
+
+import (
+	"flag"
+	"os"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util/validation"
+
+	"gopkg.in/yaml.v2"
+)
+
+// Limits describe all the limits for users; can be used to describe global default
+// limits via flags, or per-user limits via yaml config.
+type Limits struct {
+	// Distributor enforced limits.
+	IngestionRate          float64       `yaml:"ingestion_rate"`
+	IngestionBurstSize     float64       `yaml:"ingestion_burst_size"`
+	MaxLabelNameLength     int           `yaml:"max_label_name_length"`
+	MaxLabelValueLength    int           `yaml:"max_label_value_length"`
+	MaxLabelNamesPerSeries int           `yaml:"max_label_names_per_series"`
+	RejectOldSamples       bool          `yaml:"reject_old_samples"`
+	RejectOldSamplesMaxAge time.Duration `yaml:"reject_old_samples_max_age"`
+	CreationGracePeriod    time.Duration `yaml:"creation_grace_period"`
+	EnforceMetricName      bool          `yaml:"enforce_metric_name"`
+
+	// Ingester enforced limits.
+	MaxStreamsPerUser int `yaml:"max_streams_per_user"`
+
+	// Querier enforced limits.
+	MaxChunksPerQuery          int           `yaml:"max_chunks_per_query"`
+	MaxQueryLength             time.Duration `yaml:"max_query_length"`
+	MaxQueryParallelism        int           `yaml:"max_query_parallelism"`
+	CardinalityLimit           int           `yaml:"cardinality_limit"`
+	MaxStreamsMatchersPerQuery int           `yaml:"max_streams_matchers_per_query"`
+
+	// Config for overrides, convenient if it goes here.
+	PerTenantOverrideConfig string        `yaml:"per_tenant_override_config"`
+	PerTenantOverridePeriod time.Duration `yaml:"per_tenant_override_period"`
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (l *Limits) RegisterFlags(f *flag.FlagSet) {
+	f.Float64Var(&l.IngestionRate, "distributor.ingestion-rate-limit", 1, "Per-user ingestion rate limit in sample size per second.")
+	f.Float64Var(&l.IngestionBurstSize, "distributor.ingestion-burst-size", 1.5, "Per-user allowed ingestion burst size (in sample size). Warning, very high limits will be reset every -distributor.limiter-reload-period.")
```

Comment:
I'm confused with the units of our rate limit, I think we should set it up to be bytes/sec?

Reply:
Sorry, I forgot to put the unit. It is currently MB/s. I was going to discuss this tomorrow in weekly call. I think bytes/sec would be too granular, maybe KB/s?

Reply:
yeah, i was debating the same, I went with bytes/s because it covers anybody's use case if for some reason they really wanted stringent limits.... but you are correct it would be very granular... I agree, lets bring it up tomorrow